pkg/apis/core/v1/helper/helpers.go

```diff
@@ -25,59 +25,15 @@ import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/selection"
-	"k8s.io/apimachinery/pkg/util/validation"
+	corev1helpers "k8s.io/component-helpers/scheduling/corev1"
 	"k8s.io/kubernetes/pkg/apis/core/helper"
 )
 
-// IsExtendedResourceName returns true if:
-// 1. the resource name is not in the default namespace;
-// 2. resource name does not have "requests." prefix,
-// to avoid confusion with the convention in quota
-// 3. it satisfies the rules in IsQualifiedName() after converted into quota resource name
-func IsExtendedResourceName(name v1.ResourceName) bool {
```

Comment:
these are used in API validation, and changing them has significant compatibility implications. moving them to a helpers package that is not under API approval is problematic. how can we ensure that they do not change behavior (either by allowing new values, or by disallowing old values) accidentally in ways that will break us?

Reply:
We might move all the helper functions under k8s.io/component-helpers/api, where API approval will be required. Anything under `api` directory will need to go through a compatibility check as well. Would that be sufficient?

Reply:
That could help make sure changes were properly reviewed.

If we expose these externally, they will start getting used by out of tree components using versions of the component-helpers library which will lag the API server. Those components will think they are using identical logic to the API server, but will be missing additions made to the helper functions. That will cause those components to consider some data invalid which the API server considers valid. That seems like a problem we don't have today that this move would introduce.

Reply:
Good point. I don't have a resolution for it right now. We can continue the discussion into 1.21 frame. No need to rush it to resolve this for 1.20.

Reply:
Helpers used directly in validation:
- IsNativeResourceName (pkg/apis/core/v1/validation/validation.go)
- IsExtendedResourceName (pkg/scheduler/apis/config/validation/validation.go, pkg/apis/core/v1/validation/validation.go)
- IsHugePageResourceName (pkg/apis/core/validation/validation.go)

IIUC, this basically boils down to out of tree components using different version of component-helper than the one used by API server. So newer version of k8s (and thus of component-helpers) might accept new fields which are invalid in older versions (or the other way around when an old field is no longer accepted). Which basically forces out of tree components to have the same version of component-helpers as version of API server they are talking to.

> That seems like a problem we don't have today that this move would introduce.

Given the helpers moved in this PR are used in API server validation and by kube-scheduler (`IsExtendedResourceName` helper at least), it's already possible to have a helper with more relaxed constraint in version N (e.g. adding a new resource namespace or syntax) than in version N-1. So, when an API server of version N-1 and kube-scheduler of version N is ran, improper reviews of kube-scheduler PRs might create a case where kube-scheduler accepts a string (e.g. resource name) which is considered valid in N but invalid in N-1 (e.g. by not checking which version of API server the scheduler is talking to).

Some out of tree components might have relaxed acceptance for backward compatibility and use the helpers with assumption they are compatible with a range of API server versions. In which case, it's important to stress in k8s.io/component-helpers/api README (or docs) to keep this case in mind and make sure the right version of API server and out of tree component are compatible with helpers as well.

> how can we ensure that they do not change behavior (either by allowing new values, or by disallowing old values) accidentally in ways that will break us

Can't think of anything else but strict reviews and proper CI.

Reply:
Maybe just stress and make sure that importing k8s.io/component-helpers/api is only meant for core components and any other component will be importing it at its own risk?

Reply:
Citing the repo's [README](https://github.com/kubernetes/component-helpers/blob/master/README.md)

> This repository provides helpers primarily for core components (core components as described in Create a k8s.io/component-base repo) which are required by at least two separate binaries in kubernetes org. Yet, still with a high level of abstraction.

So I don't think this should be a concern for now.